/*
 * Copyright 2026 Muteb Almughyiri
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.testin.editor;

import com.intellij.openapi.fileEditor.FileEditor;
import com.intellij.openapi.fileEditor.FileEditorPolicy;
import com.intellij.openapi.fileEditor.FileEditorProvider;
import com.intellij.openapi.fileTypes.FileType;
import com.intellij.openapi.project.DumbAware;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import org.jetbrains.annotations.NotNull;

public class UnifiedEditorProvider implements FileEditorProvider, DumbAware {

    @Override
    public boolean accept(final @NotNull Project p, final @NotNull VirtualFile file) {
        return file instanceof UnifiedVirtualFile vf && vf.isValid();
    }

    @Override
    public @NotNull FileEditor createEditor(final @NotNull Project p, final @NotNull VirtualFile file) {
        if (file instanceof UnifiedVirtualFile unifiedFile) {

            final @NotNull FileType ft = unifiedFile.getFileType();
            if (!(ft instanceof EditorType editorType))
                throw new IllegalArgumentException("Unknown FileType: " + ft);

            final @NotNull TestinEditor editor = editorType.getFactory().apply(p, unifiedFile);
            return new UnifiedFileEditor(p, unifiedFile, editor);
        }

        throw new IllegalArgumentException("Unsupported virtual file type: " + file.getClass().getName());
    }

    @Override
    public @NotNull String getEditorTypeId() {
        return "test-git-unified-editor";
    }

    @Override
    public @NotNull FileEditorPolicy getPolicy() {
        return FileEditorPolicy.HIDE_DEFAULT_EDITOR;
    }
}